### PR TITLE
[CDC toolkit] : Bootstrapping mysql cdc functionality

### PR DESCRIPTION
**Pull request by akashkulk**
List of features to punt on (not handled in this PR) : https://github.com/airbytehq/airbyte-internal-issues/issues/9764 

Initial pass at CDC happy path MVP: 
- Starting debezium with previous state + no state

Code structure : 
- `JdbcPartionsCreatorFactory` : entry point to create CdcPartitionCreator
- `CdcContext` : bean holding all interfaces required for CDC. `DebeziumManager`, `CdcPositionMapper`, `DebeziumEventConverter`, `InitialCdcStateCreatorFactory`. For some of these interfaces, Mysql specific logic is implemented on the pojo side, including interpreting `OpaqueStateValue`
- Much of Rodi's work around scheduling is merged into this PR, but included here : https://github.com/airbytehq/airbyte/pull/45609/files. Currently, the model for simplicity is (i) CDC partition reader is run exactly once in a sync and (ii) While `CDCPartiionReader` runs, no other partitions can run

Currently, this has been tested via `CliRunner.source()` by pointing it to a test db. More unit tests + integration tests forthcoming

### CODE REVIEW

File: airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/cdc/MySqlDebeziumEventConverter.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        val after: JsonNode = record.after()
+        val source: JsonNode = record.source()
+
+        // Value of before and after may be a null or a NullNode object,
+        // representing a "null" in json
+        val baseNode =
+            when (after.isNull == true) {
+                true -> before
+                false -> after
+            }
```
Comment (postamar) [inline airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/cdc/MySqlDebeziumEventConverter.kt:30]: FYI this might read better as a simple if-else expression

File: airbyte-integrations/connectors/source-mysql-v2/src/test/kotlin/io/airbyte/integrations/source/mysql/MysqlCdcIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    @Property(name = "airbyte.connector.config.json", value = CDC_CONFIG_JSON)
+    @Property(name = "airbyte.connector.catalog.json", value = CDC_CATALOG)
+    fun testReadNoState() {
+        val pojoConfig: MysqlSourceConfigurationJsonObject = supplier.get()
+        val configuredCatalog = catalogFactory.make(CDC_CATALOG)
+        val output: BufferingOutputConsumer =
+            CliRunner.source("read", pojoConfig, configuredCatalog).run()
+        Assertions.assertNotNull(output.specs())
+    }
+}
```
Comment (postamar) [inline airbyte-integrations/connectors/source-mysql-v2/src/test/kotlin/io/airbyte/integrations/source/mysql/MysqlCdcIntegrationTest.kt:64]: yay Micronaut! This is a good start. We'll flesh it out as it goes.

File: airbyte-integrations/connectors/source-mysql-v2/src/test/kotlin/io/airbyte/integrations/source/mysql/MysqlCdcIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    @Property(name = "airbyte.connector.config.json", value = CDC_CONFIG_JSON)
+    @Property(name = "airbyte.connector.catalog.json", value = CDC_CATALOG)
+    fun testReadWithState() {
+        val pojoConfig: MysqlSourceConfigurationJsonObject = supplier.get()
+        val configuredCatalog = catalogFactory.make(CDC_CATALOG)
+        val list: List<io.airbyte.protocol.models.v0.AirbyteStateMessage>? =
+            ValidatedJsonUtils.parseList(
+                    io.airbyte.protocol.models.v0.AirbyteStateMessage::class.java,
+                    AIRBYTE_STATE_MESSAGE_JSON_2
+                )
```
Comment (postamar) [inline airbyte-integrations/connectors/source-mysql-v2/src/test/kotlin/io/airbyte/integrations/source/mysql/MysqlCdcIntegrationTest.kt:40]: You can inject this using `@Property` I think?

**Review by postamar — COMMENTED** [review]
I did a first pass reviewing this. This is a good start!

Comments starting with FYI are informative only, I don't mean that action needs to be taken, they're for little things and we have bigger fish to fry.

I do have a few immediate requests but for the time being they're mostly requests for clarification.

Here's what I think we should do as next steps, let me know what you think:
1. Figure out how to deal with the dependencies to decouple the toolkits. 
2. Clean this PR up enough so that it passes CI
3. Merge it, at that point the WIP areas of the code will clearly be only in bulk-cdk-toolkit-extract-cdc and source-mysql-v2
4. We keep on hacking on those.

I think the dependencies thing needs to be solved soon. The idea is that mongo should not pull in JDBC dependencies, and likewise oracle should not pull in cdc dependencies (even though it eventually will, of course). I'm keen to not turn the jdbc toolkit into a construction area while we work on CDC, though it's obvious that we need to make changes to connect its machinery to CDC. Rodi's changes in this PR show that it can be done unintrusively enough, which is nice. I'm happy to take on this task, I have reasonable hopes to be done before EOW.

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        HEARTBEAT_NOT_PROGRESSING
+    }
+
+    private fun requestClose(closeLogMessage: String, closeReason: DebeziumCloseReason) {
+        log.info { closeLogMessage }
+        log.info { closeReason }
+        // TODO : send close analytics message
+        // TODO : Close the engine. Note that engine must be closed in a different thread
+        // than the running engine.
+        CoroutineScope(Dispatchers.IO).launch { engine?.close() }
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt:201]: Yes. This is how I understand this as well.
Because debezium never `suspend`s it will remain on its thread for lifetime.
So `launch`ing something is guaranteed to run on a different thread.

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/AirbyteFileOffsetBackingStore.kt [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+private val LOGGER = KotlinLogging.logger {}
+/**
+ * This class handles reading and writing a debezium offset file. In many cases it is duplicating
+ * logic in debezium because that logic is not exposed in the public API. We mostly treat the
+ * contents of this state file like a black box. We know it is a Map&lt;ByteBuffer, Bytebuffer&gt;.
+ * We deserialize it to a Map&lt;String, String&gt; so that the state file can be human readable. If
+ * we ever discover that any of the contents of these offset files is not string serializable we
+ * will likely have to drop the human readability support and just base64 encode it.
+ */
+class AirbyteFileOffsetBackingStore(
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/AirbyteFileOffsetBackingStore.kt:33]: Yes, I'll add comments for logic that I've copied over

File: airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    implementation project(':airbyte-cdk:bulk:core:bulk-cdk-core-extract')
+    implementation 'io.debezium:debezium-api:2.4.0.Final'
+    implementation 'io.debezium:debezium-embedded:2.4.0.Final'
+    implementation project(':airbyte-cdk:java:airbyte-cdk:airbyte-cdk-dependencies')
+    implementation 'io.debezium:debezium-embedded:2.7.1.Final'
+
+    testFixturesApi testFixtures(project(':airbyte-cdk:bulk:core:bulk-cdk-core-base'))
+    testFixturesApi testFixtures(project(':airbyte-cdk:bulk:core:bulk-cdk-core-extract'))
+    testFixturesImplementation 'com.h2database:h2:2.2.224'
+    testFixturesImplementation 'org.apache.commons:commons-lang3:3.14.0'
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle:12]: Removed

File: airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle [hunk 7]
```diff
@@ -0,0 +1,13 @@
+dependencies {
+    implementation project(':airbyte-cdk:bulk:core:bulk-cdk-core-base')
+    implementation project(':airbyte-cdk:bulk:core:bulk-cdk-core-extract')
+    implementation 'io.debezium:debezium-api:2.4.0.Final'
+    implementation 'io.debezium:debezium-embedded:2.4.0.Final'
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle:5]: Done

File: airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    implementation project(':airbyte-cdk:java:airbyte-cdk:airbyte-cdk-dependencies')
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle:6]: Let me see if I can go through these deps and copy over the ones I need

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            engine = createDebeziumEngine()
+            engine?.run()
+        }
+    }
+
+    override fun checkpoint(): PartitionReadCheckpoint {
+        if (opaqueStateValue != null) {
+            return PartitionReadCheckpoint(propertyManager.readOffsetState(), numRecords)
+        } else {
+            return PartitionReadCheckpoint(initialCdcStateCreatorFactory.make(), 0)
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt:76]: The desired behavior here is to : 
1. Skip running the dbz engine if it's the first run (no prev debezium state)
2. Return initial debezium state as the checkpoint as that is the point in the binlog the sync will resume from on subsequent runs

With that in mind, this should be easy to move to this logic to run

[the remark this answers is not in the document]
Comment (postamar) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt:76]: I must admit, I don't understand the reasoning behind point (2). I look forward to discussing synchronously.

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+    fun createDebeziumEngine(): DebeziumEngine<ChangeEvent<String?, String?>>? {
+        log.info {
+            "Using DBZ version: ${DebeziumEngine::class.java.getPackage().implementationVersion}"
+        }
+        return DebeziumEngine.create(Json::class.java)
+            .using(propertyManager.getPropertiesForSync(opaqueStateValue))
+            .using(OffsetCommitPolicy.AlwaysCommitOffsetPolicy())
+            .notifying { event: ChangeEvent<String?, String?> ->
+                if (event.value() == null) {
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt:94]: They are truly worthless. Copying over this comment from legacy CDK : 

  // debezium outputs a tombstone event that has a value of null. this is an
  // artifact of how it
  // interacts with kafka. we want to ignore it.
  // more on the tombstone:
  // https://debezium.io/documentation/reference/2.2/transformations/event-flattening.html

**Comment by rodireich** [comment]
Regarding the dependency between toolkits
Here's the rationale:
We intend all JDBC connectors to also have CDC.
MongoDB is the exception in the sense that it is CDC but not JDBC.

For this reason the JDBC toolkit depends on CDC and has a small entry in its `PartitionsCreatorFactory` which looks at `configuration.global` and routes partitions creators accordingly - that is running a CDC partition creator and making the JDBC partitions be `CdcAware`. But only in the case of global.

On the other hand the CDC toolkit doesn't know about jdbc and in Mongo's application yaml we're going to have an
```
extract:
  cdc:
```
which will only inject CDC classes.


tl;dr: jdbc connector is also cdc, but cdc is not necessarily jdbc.

File: airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle (hunk 8 above)
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/build.gradle:6]: Done

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/Partitions.kt [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        return false
+    }
+
+    @Synchronized fun unlock() = mutex.unlock()
+
+    val isLocked: Boolean
+        get() = mutex.isLocked
+}
+
+interface cdcResourceTaker {}
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/Partitions.kt]: Added comment

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/Partitions.kt [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.concurrent.atomic.AtomicInteger
+import kotlinx.coroutines.sync.Mutex
+
+/** A [Mutex] wrapper that only allows locking n times */
+internal class LimitedLockMutex(maxCount: Int) {
+    private val mutex = Mutex()
+    private var locksLeft = AtomicInteger(maxCount)
+
+    @Synchronized
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/Partitions.kt:15]: `@Synchronized` annotation in kotlin generates a `synchronized` jvm function - it look like it maps directly.
Synchronized(this) { block } is similar to java's synchronized block I think.

[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/Partitions.kt:15]: because locksLeft and mutex are thread safe they don't need protection themselves, I'm trying to protect from a situation one thread is check `canLock()` while another thread is in the middle of `tryLock()`.

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionCreator.kt [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicReference
+
+class CdcPartitionCreator<
+    A : CdcSharedState,
+>(val sharedState: A, cdcContext: CdcContext, opaqueStateValue: OpaqueStateValue?) :
+    PartitionsCreator, CdcAware {
+    private val acquiredResources = AtomicReference<AcquiredResources>()
+    val cdcContext = cdcContext
+    val opaqueStateValue = opaqueStateValue
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionCreator.kt:17]: Done

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+                    return@notifying
+                }
+                val record = DebeziumRecord(Jsons.deserialize(event.value()))
+                // TODO : Migrate over all of the timeout/heartbeat timeout logic
+                if (!record.isHeartbeat) {
+                    numRecords++
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt:111]: I don't think so, this partition will only be run in a single thread. Our resource acquisition phase also ensures that only the CDC partition is only run once

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcSourceConfiguration.kt [hunk 15]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.read
+
+import io.airbyte.cdk.command.SourceConfiguration
+
+interface CdcSourceConfiguration : SourceConfiguration {}
```
[the remark this answers is not in the document]
Comment (akashkulk) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcSourceConfiguration.kt:9]: That's correct - I've added a comment here

File: airbyte-integrations/connectors/source-mysql-v2/src/test/kotlin/io/airbyte/integrations/source/mysql/MysqlCdcIntegrationTest.kt (hunk 3 above)
Comment (akashkulk) [reply airbyte-integrations/connectors/source-mysql-v2/src/test/kotlin/io/airbyte/integrations/source/mysql/MysqlCdcIntegrationTest.kt:40]: This test will soon go away - mainly used for the iteration loop. Rodi is working on another integration test here

File: gradle/wrapper/gradle-wrapper.properties [hunk 16]
```diff
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.6-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.9-bin.zip
```
[the remark this answers is not in the document]
Comment (rodireich) [reply gradle/wrapper/gradle-wrapper.properties]: Sent by mistake from my environment. Rolled back

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+import org.apache.kafka.connect.source.SourceRecord
+
+class CdcPartitionReader<S : CdcSharedState>(
+    private val sharedState: S,
+    private val cdcContext: CdcContext,
+    opaqueStateValue: OpaqueStateValue?,
+) : PartitionReader, CdcAware, cdcResourceTaker {
+
+    private val log = KotlinLogging.logger {}
+    private var engine: DebeziumEngine<ChangeEvent<String?, String?>>? = null
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcPartitionReader.kt:40]: Updated 
I never quite understood `lateinit`. are you supposed to check with `.isInitialized`? isn't that nullable in different words?)

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/DefaultCdcSharedState.kt [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package io.airbyte.cdk.read
+
+import io.airbyte.cdk.cdc.CdcPartitionReader
+import io.airbyte.cdk.command.SourceConfiguration
+import kotlinx.coroutines.sync.Semaphore
+
+class DefaultCdcSharedState(override val configuration: SourceConfiguration) : CdcSharedState {
+    internal val semaphore = Semaphore(configuration.maxConcurrency)
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/DefaultCdcSharedState.kt]: That's a very good point. they should be acquiring the same resource. I changed this to not create a copy of resources. 
Thanks for noticing!

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcSharedState.kt [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.read
+
+import io.airbyte.cdk.cdc.CdcPartitionReader
+import io.airbyte.cdk.command.SourceConfiguration
+
+interface CdcSharedState {
+    val configuration: SourceConfiguration
```
[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io.airbyte.cdk/read/CdcSharedState.kt]: removed configuration as it's no longer used

**Review by postamar — APPROVED** [review]
LGTM assuming https://github.com/airbytehq/airbyte/pull/45706 gets merged into this branch first. If there are lingering issues in my PR (it's very possible!) you can always fix them here and then merge, no need to keep me in the loop if it's not necessary.

There's lots to be done but this PR enables us to improve the cdc toolkit and source-mysql-v2 incrementally without messing with the rest of the CDK, so this is good enough in my book.
